From bf142234378944289066aca4cb12b26cb21afa09 Mon Sep 17 00:00:00 2001
From: Hien Huynh <hien.huynh.px@renesas.com>
Date: Tue, 27 Oct 2020 11:30:40 +0700
Subject: [PATCH 04/18] rzg: drivers: add board identification

This commit adds driver to identify board revision for rzg gen2
platform.

Signed-off-by: Hien Huynh <hien.huynh.px@renesas.com>
---
 drivers/renesas/rzg/board/board.c | 118 ++++++++++++++++++++++++++++++++++++++
 drivers/renesas/rzg/board/board.h |  32 +++++++++++
 2 files changed, 150 insertions(+)
 create mode 100644 drivers/renesas/rzg/board/board.c
 create mode 100644 drivers/renesas/rzg/board/board.h

diff --git a/drivers/renesas/rzg/board/board.c b/drivers/renesas/rzg/board/board.c
new file mode 100644
index 0000000..d8d7eb0
--- /dev/null
+++ b/drivers/renesas/rzg/board/board.c
@@ -0,0 +1,118 @@
+/*
+ * Copyright (c) 2020, Renesas Electronics Corporation. All rights
+ * reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#include <stdint.h>
+
+#include <lib/utils_def.h>
+
+#include <iic_dvfs.h>
+#include <lib/mmio.h>
+#include "rzg_def.h"
+
+#include "board.h"
+
+#ifndef BOARD_DEFAULT
+#if (RZG_LSI == RZG_G2E)
+#define BOARD_DEFAULT		(BOARD_EK874 << BOARD_CODE_SHIFT)
+#elif (RZG_LSI == RZG_G2N)
+#define BOARD_DEFAULT		(BOARD_HIHOPE_RZG2N << BOARD_CODE_SHIFT)
+#elif (RZG_LSI == RZG_G2H)
+#define BOARD_DEFAULT		(BOARD_HIHOPE_RZG2H << BOARD_CODE_SHIFT)
+#else
+#define BOARD_DEFAULT		(BOARD_HIHOPE_RZG2M << BOARD_CODE_SHIFT)
+#endif
+#endif
+
+#define BOARD_CODE_MASK		(0xF8)
+#define BOARD_REV_MASK		(0x07)
+#define BOARD_CODE_SHIFT	(0x03)
+#define BOARD_ID_UNKNOWN	(0xFF)
+
+
+#define	GPIO_INDT5		0xE605500C
+#define	GP5_19_BIT		(0x01 << 19)
+#define	GP5_21_BIT		(0x01 << 21)
+#define	GP5_25_BIT		(0x01 << 25)
+
+#define HM_ID	{ 0x10U, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU }
+#define HN_ID	{ 0x20U, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU }
+#define HH_ID	{ 0x40U, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU }
+#define EK_ID	{ 0x10U, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU, 0xFFU }
+
+const char *g_board_tbl[] = {
+	[BOARD_HIHOPE_RZG2M]	= "HiHope RZ/G2M",
+	[BOARD_HIHOPE_RZG2N]	= "HiHope RZ/G2N",
+	[BOARD_HIHOPE_RZG2H]	= "HiHope RZ/G2H",
+	[BOARD_EK874]		= "EK874 RZ/G2E",
+	[BOARD_UNKNOWN]		= "unknown"
+};
+
+int32_t rzg_get_board_type(uint32_t *type, uint32_t *rev)
+{
+	int32_t ret = 0;
+	const uint8_t board_tbl[][8] = {
+		[BOARD_HIHOPE_RZG2M] = HM_ID,
+		[BOARD_HIHOPE_RZG2N] = HN_ID,
+		[BOARD_HIHOPE_RZG2H] = HH_ID,
+		[BOARD_EK874] = EK_ID,
+	};
+	static uint8_t board_id = BOARD_ID_UNKNOWN;
+#if (RZG_LSI == RZG_G2N) | (RZG_LSI == RZG_G2M)
+	uint32_t reg, boardInfo, read_rev;
+#elif (RZG_LSI == RZG_G2H)
+	uint32_t boardInfo;
+#else
+	uint32_t read_rev;
+#endif
+
+	if (board_id != BOARD_ID_UNKNOWN)
+		goto get_type;
+
+	board_id = BOARD_DEFAULT;
+
+get_type:
+	*type = ((uint32_t) board_id & BOARD_CODE_MASK) >> BOARD_CODE_SHIFT;
+
+	if (*type >= ARRAY_SIZE(board_tbl)) {
+		/* no revision information, set Rev0.0. */
+		*rev = 0;
+		return ret;
+	}
+
+#if (RZG_LSI == RZG_G2M)
+	reg = mmio_read_32(RZG_PRR);
+	if (RZG_G2M_CUT_VER11 == (reg & PRR_CUT_MASK))
+	{
+		read_rev = (uint8_t)(board_id & BOARD_REV_MASK);
+		*rev = board_tbl[*type][read_rev];
+	}
+	else
+	{
+		boardInfo = mmio_read_32(GPIO_INDT5) & (GP5_19_BIT |GP5_21_BIT);
+		*rev = (((boardInfo & GP5_19_BIT) >> 14) | ((boardInfo & GP5_21_BIT) >> 17)) + 0x30;
+	}
+#elif (RZG_LSI == RZG_G2N)
+	reg = mmio_read_32(GPIO_INDT5);
+	if (reg & GP5_25_BIT)
+	{
+		read_rev = (uint8_t)(board_id & BOARD_REV_MASK);
+		*rev = board_tbl[*type][read_rev];
+	}
+	else
+	{
+		boardInfo = reg & (GP5_19_BIT |GP5_21_BIT);
+		*rev = (((boardInfo & GP5_19_BIT) >> 14) | ((boardInfo & GP5_21_BIT) >> 17)) + 0x30;
+	}
+#elif (RZG_LSI == RZG_G2H)
+	boardInfo = mmio_read_32(GPIO_INDT5) & (GP5_19_BIT |GP5_21_BIT);
+	*rev = (((boardInfo & GP5_19_BIT) >> 14) | ((boardInfo & GP5_21_BIT) >> 17)) + 0x30;
+#else
+	read_rev = (uint8_t)(board_id & BOARD_REV_MASK);
+	*rev = board_tbl[*type][read_rev];
+#endif
+	return ret;
+}
diff --git a/drivers/renesas/rzg/board/board.h b/drivers/renesas/rzg/board/board.h
new file mode 100644
index 0000000..e8c0e9f
--- /dev/null
+++ b/drivers/renesas/rzg/board/board.h
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) 2020, Renesas Electronics Corporation. All rights
+ * reserved.
+ *
+ * SPDX-License-Identifier: BSD-3-Clause
+ */
+
+#ifndef BOARD_H
+#define BOARD_H
+
+#define BOARD_EK874			(0x00U)
+#define BOARD_HIHOPE_RZG2M		(0x01U)
+#define BOARD_HIHOPE_RZG2N		(0x02U)
+#define BOARD_HIHOPE_RZG2H		(0x03U)
+#define BOARD_UNKNOWN			(BOARD_HIHOPE_RZG2H + 1U)
+
+#define BOARD_REV_UNKNOWN		(0xFF)
+
+extern const char *g_board_tbl[];
+
+/************************************************************************
+ * Revisions are expressed in 8 bits.
+ *  The upper 4 bits are major version.
+ *  The lower 4 bits are minor version.
+ ************************************************************************/
+#define GET_BOARD_MAJOR(a)	((uint32_t)(a) >> 0x4)
+#define GET_BOARD_MINOR(a)	((uint32_t)(a) &  0xF)
+#define GET_BOARD_NAME(a)	(g_board_tbl[(a)])
+
+int32_t rzg_get_board_type(uint32_t *type, uint32_t *rev);
+
+#endif /* BOARD_H */
-- 
2.7.4

